docs/tutorial/reduction.rst

```diff
@@ -0,0 +1,720 @@
+        T* back,
+        F op,
+        T zero_elem,
+        uint32_t front_size)
+    {
+        // ...
+    }
+
+The kernel signature and the reduction factor are the same as in previous cases; only the implementation differs.
+
+.. code-block:: C++
+
+    static constexpr uint32_t WarpCount = BlockSize / WarpSize;
+
+    __shared__ T shared[WarpCount];
+
+    auto read_global_safe =
+        [&](const uint32_t i) { return i < front_size ? front[i] : zero_elem; };
+    auto read_shared_safe =
+        [&](const uint32_t i) { return i < WarpCount ? shared[i] : zero_elem; };
+
+    const uint32_t tid = threadIdx.x,
+                   bid = blockIdx.x,
+                   gid = bid * (blockDim.x * 2) + tid,
+                   wid = tid / WarpSize,
+                   lid = tid % WarpSize;
+
+    // Read input from front buffer to local
+    T res = op(read_global_safe(gid), read_global_safe(gid + blockDim.x));
+
+As we communicate the results of warps through shared memory, same number of elements are required in the shared memory as warps within the block. Similar to how you can only launch kernels at block granularity, you can only warp reduce with ``WarpSize`` granularity due to the collective nature of the cross-lane builtins. To address this, you can use ``read_shared_safe`` to pad overindexing by reading ``zero_elem``. Reading from global remains unaffected.
+
+.. code-block:: C++
+
+    // Perform warp reductions and communicate results via shared
+    // for (uint32_t ActiveWarps = WarpCount;
+    //      ActiveWarps != 0;
+    //      ActiveWarps = ActiveWarps != 1 ?
+    //          divide_ceil(ActiveWarps, WarpSize) :
+    //          ActiveWarps = 0)
+    tmp::static_for<
+        WarpCount,
+        tmp::not_equal<0>,
+        tmp::select<
+            tmp::not_equal<1>,
+            tmp::divide_ceil<WarpSize>,
+            tmp::constant<0>>>([&]<uint32_t ActiveWarps>()
+    {
+        if(wid < ActiveWarps)
+        {
+            // Warp reduction
+            tmp::static_for<WarpSize / 2, tmp::not_equal<0>, tmp::divide<2>>([&]<int Delta>()
+            {
+                res = op(res, __shfl_down(res, Delta));
+            });
+
+            // Write warp result from local to shared
+            if(lid == 0)
+                shared[wid] = res;
+        }
+        __syncthreads();
+
+        // Read warp result from shared to local
+        res = read_shared_safe(tid);
+    });
+
+    // Write result from local to back buffer
+    if(tid == 0)
+        back[bid] = res;
+
+``ActiveWarps`` iterates from ``WarpCount`` until it reaches ``0``. Every iteration of ``ActiveWarps`` reduces the ``WarpSize``. In cases where the partial result count isn't a divisor of ``ActiveWarps`` and you need to launch an extra warp, use ``tmp::divide_ceil``, which always rounds to positive infinity. You need the tertiary ``tmp::select``, because such division never reaches ``0``, so you must terminate the loop after the last warp concludes.
```

Comment:
```suggestion
``ActiveWarps`` iterates from ``WarpCount`` until it reaches ``0``. Every iteration of ``ActiveWarps`` reduces the ``WarpSize``. In cases where the partial result count isn't a divisor of ``ActiveWarps`` and you need to launch an extra warp, use ``tmp::divide_ceil``, which always rounds to positive infinity. The tertiary ``tmp::select`` is required because such division never reaches ``0``, so you must terminate the loop after the last warp concludes.
```